### ISSUE TITLE
loss become inf , then Nan

### ISSUE THREAD

**Issue by XiaogangYin** [zhiqwang/image-captioning#13]
mtwi_2018_train/images/001807_00031.jpg
>> Train: [1][108/90000]        Time 0.348 (0.361)      Data 0.003 (0.006)      Loss 30.0584 (31.5477)
mtwi_2018_train/images/007947_00013.jpg
>> Train: [1][109/90000]        Time 0.422 (0.361)      Data 0.003 (0.006)      Loss inf (inf)
mtwi_2018_train/images/002394_00012.jpg
>> Train: [1][110/90000]        Time 0.332 (0.361)      Data 0.003 (0.006)      Loss nan (nan)
------
command:
python ./main.py --dataset-root mtwi_2018_train --arch densenet121 --alphabet ./data/alphabet_decode_5990.txt --lr 1e-6 --optimizer rmsprop --gpu-id -1 --workers 1 --not-pretrained --batch-size 1 --keep-ratio --print-freq 1

-----
attach 007947_00013.jpg 
![image](https://user-images.githubusercontent.com/8209028/57348086-56112a80-7187-11e9-8061-61ded3ef38b3.png)

**Comment by zhiqwang** [comment]
I didn't know the alphabet is suitable for this datasets?

Recently, I will download this datasets, and verify this CRNN repo.

**Comment by zhiqwang** [comment]
And the current Transform used for `keep-ratio` suppose all image's width is bigger than height, the attached image is fail for this condition.  Refer issue #11 for more detail.

**Comment by USTC-Keyanjie** [comment]
I have encountered the same problem!

>> Creating directory if it does not exist:
>> './checkpoint/densenet121_rmsprop_lr5.0e-05_wd5.0e-04_bsize64_height32_keep_ratio'
>> Using model from scratch (random weights) 'densenet121'
>> Train: [1][1/10]	Time 80.764 (80.764)	Data 0.220 (0.220)	Loss 4.1266 (4.1266)
>> Train: [1][10/10]	Time 74.218 (75.728)	Data 0.005 (0.025)	Loss 3.2840 (3.6721)
>> Val: [1][1/2]	Time 17.361 (17.361)	Accu 0.000
>> Val: [1][2/2]	Time 5.320 (11.340)	Accu 0.000
>> Train: [2][1/10]	Time 77.894 (77.894)	Data 0.195 (0.195)	Loss 3.2516 (3.2516)
>> Train: [2][10/10]	Time 79.049 (76.578)	Data 0.005 (0.024)	Loss 2.8698 (3.0496)
>> Val: [2][1/2]	Time 17.272 (17.272)	Accu 0.000
>> Val: [2][2/2]	Time 5.313 (11.292)	Accu 0.000
>> Train: [3][1/10]	Time 75.839 (75.839)	Data 0.210 (0.210)	Loss 2.8418 (2.8418)
>> Train: [3][10/10]	Time 74.184 (76.081)	Data 0.005 (0.026)	Loss inf (inf)
>> Val: [3][1/2]	Time 17.391 (17.391)	Accu 0.000
>> Val: [3][2/2]	Time 5.323 (11.357)	Accu 0.000
>> Train: [4][1/10]	Time 63.459 (63.459)	Data 0.209 (0.209)	Loss nan (nan)
>> Train: [4][10/10]	Time 66.387 (65.531)	Data 0.005 (0.026)	Loss nan (nan)
>> Val: [4][1/2]	Time 17.400 (17.400)	Accu 0.000
>> Val: [4][2/2]	Time 5.370 (11.385)	Accu 0.000
>> Train: [5][1/10]	Time 67.657 (67.657)	Data 0.226 (0.226)	Loss nan (nan)
>> Train: [5][10/10]	Time 65.224 (65.783)	Data 0.005 (0.027)	Loss nan (nan)
>> Val: [5][1/2]	Time 17.329 (17.329)	Accu 0.000
>> Val: [5][2/2]	Time 5.386 (11.357)	Accu 0.000
>> Train: [6][1/10]	Time 67.789 (67.789)	Data 0.215 (0.215)	Loss nan (nan)
>> Train: [6][10/10]	Time 65.523 (65.778)	Data 0.006 (0.025)	Loss nan (nan)
>> Val: [6][1/2]	Time 17.458 (17.458)	Accu 0.000
>> Val: [6][2/2]	Time 5.400 (11.429)	Accu 0.000
>> Train: [7][1/10]	Time 66.216 (66.216)	Data 0.210 (0.210)	Loss nan (nan)
>> Train: [7][10/10]	Time 110.504 (78.189)	Data 0.009 (0.026)	Loss nan (nan)
>> Val: [7][1/2]	Time 29.770 (29.770)	Accu 0.000
>> Val: [7][2/2]	Time 9.196 (19.483)	Accu 0.000
>> Train: [8][1/10]	Time 112.352 (112.352)	Data 0.296 (0.296)	Loss nan (nan)

command:

python ./main.py --dataset-root ./data/cropped_images --arch densenet121 --lr 5e-5 --optimizer rmsprop --not-pretrained --keep-ratio --batch-size 64

I train this model using cropped images of bankcard in order to recognize the number in that.

![1234_5678_1234_5678@1](https://user-images.githubusercontent.com/41223233/57426869-e2891f00-7253-11e9-8ea7-53fb0f8dc775.jpg)
![4270_2000_1111_9999@0](https://user-images.githubusercontent.com/41223233/57426870-e4eb7900-7253-11e9-9fd8-225851c702be.jpg)

**Comment by zhiqwang** [comment]
@USTC-Keyanjie I think the arch `densenet_cifar` is prefer for this datasets.

**Comment by zhiqwang** [comment]
@USTC-Keyanjie Besides, I have trained a model with only numbers, and parameter `keep-ratio` performs bad, I think scale the images to fixed size is the best option. BTW, the `mean` and `std` is significant for small datasets.

**Comment by USTC-Keyanjie** [comment]
Thank you for your reply so quickly. I'm trying that.
感谢您这么快的回复。我正在尝试。

Best Regards.
顺祝时祺。

**Comment by zhiqwang** [comment]
@USTC-Keyanjie Once your model is trained, you can generate text recognition using `keep-ratio`.

**Comment by lvoegtlin** [comment]
I have the same problem that after some epochs the loss just gets nan.
The parameters I train my network with is:
```--gpu-id 0 --not-pretrained --arch densenet_cifar --dataset-root /net/dataset/voegtlil/chinese_generated_first_set/split --alphabet ./data/alphabet_decode_5990.txt --lr 5e-5```

Example file:
![00000003](https://user-images.githubusercontent.com/2383718/57453288-4b799280-7266-11e9-9ddf-1f08daa60a50.jpg)

All my images are generated and have the GT in this structure:
```00000003.jpg 1160 1460 3068 1686 5574 4533 2955 5616 5784 5828```

**Comment by zhiqwang** [comment]
@lvoegtlin It seems that this dataset is same with YCG09's? For this datasets, architecture `densenet121` is prefer.

**Comment by lvoegtlin** [comment]
@zhiqwang it is similar to YCG09. I use a ```densenet121``` but I still run into the ```nan``` problem after some epochs

**Comment by zhiqwang** [comment]
@lvoegtlin Ohh, Can you please provide a more detail description of this datasets. Actually I didn't know last six Chinese characters the example you provide, it is rarely seen.

**Comment by zhiqwang** [comment]
BTW I use the last stable version 1.1 of Pytorch to train this repo, and set `criterion = nn.CTCLoss(zero_infinity=True)` in this [line](https://github.com/zhiqwang/crnn.pytorch/blob/c2f59119a1111acb3c3cc5efc5318369a125b783/main.py#L155).

**Comment by lvoegtlin** [comment]
The dataset is just randomly generated based on ```alphabet_decode_5990.txt```. And as I don't speak Chinese I don't know what it generates and it also does not matter :)

> BTW I use the last stable version 1.1 of Pytorch to train this repo, and set `criterion = nn.CTCLoss(zero_infinity=True)` in this [line](https://github.com/zhiqwang/crnn.pytorch/blob/c2f59119a1111acb3c3cc5efc5318369a125b783/main.py#L155).

Illl try this

**Comment by zhiqwang** [comment]
@lvoegtlin I check the image you provide, the width of this image is 256, is it the same for all other images?

**Comment by lvoegtlin** [comment]
Yes, all my training images are 32*256.

**Comment by zhiqwang** [comment]
@lvoegtlin Then the option `width=256` is significant. Or simply comment this [line](https://github.com/zhiqwang/crnn.pytorch/blob/c2f59119a1111acb3c3cc5efc5318369a125b783/main.py#L136).

**Comment by zhiqwang** [comment]
Or simply comment this [line](https://github.com/zhiqwang/crnn.pytorch/blob/c2f59119a1111acb3c3cc5efc5318369a125b783/main.py#L136).

**Comment by kunyao2015** [comment]
替换了mean 和std ,所有图像resize到统一大小，训练几步就出现nan，换了几个backbone也不行，到底是什么原因呢，费解加无所适从 ,求指点  PS: pytorch 1.0.1, python3.6

**Comment by zhiqwang** [comment]
Check to master branch of pytorch, There is some bugs in the implementation of CTCLoss, More detail refet to https://github.com/pytorch/pytorch/pull/21244 . Or the `pytorch_nightly` version of PyTorch, It seems that PyTorch update `pytorch_nightly` everyday.
